### ISSUE TITLE
Reported work array size for the noncontiguous partitioner is too short

### ISSUE THREAD

**Issue by drwells** [dealii/dealii#11764]
This class states that the work array passed to `export_to_ghosted_array` et al must be "at least as large as the sum of the number of entries in the index sets passed to the constructor and the reinit() functions". I am working on a problem where there is some overlap between what I want to be put in the ghost region and what is locally owned - this works, but the computed work array size is no longer correct and AFAICT there is no way to actually get the value out of the class (I multiplied the array length by 2 to temporarily get around it).

Here is a simplified setup exhibiting the same problem:

```cpp
#include <deal.II/base/array_view.h>
#include <deal.II/base/index_set.h>
#include <deal.II/base/mpi.h>
#include <deal.II/base/mpi_noncontiguous_partitioner.h>
#include <deal.II/base/mpi_noncontiguous_partitioner.templates.h>

#include <fstream>

// Test demonstrating problem with two ranks

int main(int argc, char **argv)
{
    using namespace dealii;
    Utilities::MPI::MPI_InitFinalize mpi_initialization(argc, argv, 1);

    IndexSet local(1024);
    IndexSet ghost(1024);

    const auto rank = Utilities::MPI::this_mpi_process(MPI_COMM_WORLD);
    if (rank == 0)
      {
        local.add_range(0, 512);
        ghost.add_range(0, 1024);
      }
    else
      {
        local.add_range(512, 1024);
        ghost.add_range(0, 1024);
      }

    Utilities::MPI::NoncontiguousPartitioner partitioner(
        local, ghost, MPI_COMM_WORLD);

    // Things work correctly with the buffer stored by the object itself
    {
        std::vector<int> local_values(local.n_elements());
        std::vector<int> ghost_values(ghost.n_elements());

        partitioner.export_to_ghosted_array(
            make_const_array_view(local_values),
            make_array_view(ghost_values));
    }

    if (rank == 0)
    {
        std::cout << "Passed first check\n";
    }

    // Use the documented array length - doesn't work
    {
        std::vector<int> local_values(local.n_elements());
        std::vector<int> ghost_values(ghost.n_elements());

        const auto work_size = local.n_elements() + ghost.n_elements();

        const auto n_targets = partitioner.n_targets();
        const auto n_requests = n_targets.first + n_targets.second;

        std::vector<int> work(work_size);
        std::vector<MPI_Request> requests(n_requests);

        int tag = 1;
        partitioner.export_to_ghosted_array_start<int>(
            tag,
            make_const_array_view(local_values),
            make_array_view(work),
            requests);

        partitioner.export_to_ghosted_array_finish<int>(
            make_const_array_view(work),
            make_array_view(ghost_values),
            requests);
    }
}
```

**Comment by peterrum** [comment]
with https://github.com/dealii/dealii/pull/11765, the code can be modified to:
```
const auto n_targets = partitioner.temporary_storage_size();
```

**Comment by drwells** [comment]
Are you sure? I believe `n_targets()` returns the number of processes we need to receive from and send to, not index array sizes.

**Comment by peterrum** [comment]
Fixed^^ (`work_size`)

**Comment by drwells** [comment]
Either way, what you implemented is correct :)

**Comment by drwells** [comment]
Fixed by #11765.
